Generalize field in Circlepoint

atgrosso opened the pull request:
Depends on #98

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99)
<!-- Reviewable:end -->

andrewmilson reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99)* status: 0 of 9 files reviewed, 14 unresolved discussions (waiting on @atgrosso)

___
*[`stwo_cairo_verifier/src/circle.cairo` line 8 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EHdX88fWOihMDXfB3:-O8EHdX88fWOihMDXfB4:bzc5mga) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L8)):*
> ```Cairo
> 
> pub const M31_CIRCLE_GEN: CirclePoint<M31> =
>     CirclePoint::<M31> { x: M31 { inner: 2 }, y: M31 { inner: 1268011823 }, };
> ```

Nit: can this be implicit
___
*[`stwo_cairo_verifier/src/circle.cairo` line 31 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EI3rc691JoMGgieH6:-O8EI3rd3n9PYYvNZ2Dw:brhun0a) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L31)):*
> ```Cairo
>     // Returns the neutral element of the circle.
>     fn zero() -> CirclePoint<F> {
>         CirclePoint::<F> { x: One::<F>::one(), y: Zero::<F>::zero() }
> ```

Same here?



_Suggestion:_
```Cairo
CirclePoint { x: One::one(), y: Zero::zero() }
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 50 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EIKJZ1H0k-UYe0R86:-O8EIKJ_7JgqEx-f_zRe:b-qc19fg) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L50)):*
> ```Cairo
>         };
>         result
>     }
> ```

Why does this have to change so much? Can't it be the same as before just change scalar to u128?



_Code quote:_
```Cairo
    fn mul(
        self: @CirclePoint<F>, initial_scalar: u128
    ) -> CirclePoint<
        F
    > {
        let mut scalar = initial_scalar;
        let mut result: CirclePoint<F> = Self::zero();
        let mut cur: CirclePoint<F> = *self;
        while scalar > 0 {
            if scalar & 1 == 1 {
                result = result + cur;
            }
            cur = cur + cur;
            scalar = scalar / 2;
        };
        result
    }
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 54 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EIsTJ380jXrzjRHR6:-O8EIsTJ380jXrzjRHR7:b25hiq4) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L54)):*
> ```Cairo
> 
> impl CirclePointAdd<F, +Add<F>, +Sub<F>, +Mul<F>, +Drop<F>, +Copy<F>> of Add<CirclePoint<F>> {
>     // The operation of the circle as a group with additive notation.
> ```

I know unrelated to your changes but mind chageing to docstring here please?



_Suggestion:_
```Cairo
    /// Performs the operation of the circle as a group with additive notation.
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 56 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EIZVA2uexxDQRnuJc:-O8EIZVA2uexxDQRnuJd:brhun0a) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L56)):*
> ```Cairo
>     // The operation of the circle as a group with additive notation.
>     fn add(lhs: CirclePoint<F>, rhs: CirclePoint<F>) -> CirclePoint<F> {
>         CirclePoint::<F> { x: lhs.x * rhs.x - lhs.y * rhs.y, y: lhs.x * rhs.y + lhs.y * rhs.x }
> ```

Same here?
___
*[`stwo_cairo_verifier/src/circle.cairo` line 72 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EJLXbCprTOmaef3IL:-O8EJLXbCprTOmaef3IM:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L72)):*
> ```Cairo
>         CirclePoint { x: self.x.complex_conjugate(), y: self.y.complex_conjugate() }
>     }
> }
> ```





_Suggestion:_
```Cairo
#[generate_trait]
pub impl ComplexConjugateImpl of ComplexConjugateTrait {
    fn complex_conjugate(self: CirclePoint<QM31>) -> CirclePoint<QM31> {
        CirclePoint { x: self.x.complex_conjugate(), y: self.y.complex_conjugate() }
    }
}
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 74 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EJiWr3isom0K8QNYo:-O8EJiWr3isom0K8QNYp:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L74)):*
> ```Cairo
> }
> 
> /// Represents the coset initial + <step>.
> ```





_Suggestion:_
```Cairo
/// Represents the coset `initial + <step>`.
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 112 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EJunB8O86MI36Wd6Q:-O8EJunB8O86MI36Wd6R:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L112)):*
> ```Cairo
>     }
> 
>     fn at(self: @Coset, index: usize) -> CirclePoint::<M31> {
> ```





_Suggestion:_
```Cairo
CirclePoint<M31> 
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 122 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EK8E77ZTaqsFe1hxC:-O8EK8E77ZTaqsFe1hxD:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L122)):*
> ```Cairo
> 
>     /// Creates a coset of the form G_2n + \<G_n\>.
>     /// For example, for n=8, we get the point indices \[1,3,5,7,9,11,13,15\].
> ```





_Suggestion:_
```Cairo
    /// Creates a coset of the form `G_2n + <G_n>`.
    ///
    /// For example, for `n=8`, we get the point indices `[1,3,5,7,9,11,13,15]`.
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 124 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EKHQsC5kuhvLI3dy1:-O8EKHQsC5kuhvLI3dy2:byp18fd) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L124)):*
> ```Cairo
>     /// For example, for n=8, we get the point indices \[1,3,5,7,9,11,13,15\].
>     fn odds(log_size: u32) -> Coset {
>         //CIRCLE_LOG_ORDER
> ```

Remove
___
*[`stwo_cairo_verifier/src/circle.cairo` line 131 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EKMkjEnCGvL1K_03V:-O8EKMkjEnCGvL1K_03W:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L131)):*
> ```Cairo
>     /// Creates a coset of the form G_4n + <G_n>.
>     /// For example, for n=8, we get the point indices \[1,5,9,13,17,21,25,29\].
>     /// Its conjugate will be \[3,7,11,15,19,23,27,31\].
> ```





_Suggestion:_
```Cairo
    /// Creates a coset of the form `G_4n + <G_n>`.
    ///
    /// For example, for `n=8`, we get the point indices `[1,5,9,13,17,21,25,29]`.
    /// Its conjugate will be `[3,7,11,15,19,23,27,31]`.
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 156 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EKs_T6NmxCtSsN5zo:-O8EKs_T6NmxCtSsN5zp:b4mykql) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L156)):*
> ```Cairo
>     #[test]
>     fn test_add_1() {
>         let i = CirclePoint::<M31> { x: m31(0), y: m31(1) };
> ```

Same here? and below?



_Suggestion:_
```Cairo
        let i = CirclePoint { x: m31(0), y: m31(1) };
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 160 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EL6TJ5cyGWFtC5yqk:-O8EL6TJ5cyGWFtC5yql:bdizfvm) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L160)):*
> ```Cairo
>         let expected_result = CirclePoint::<M31> { x: -m31(1), y: m31(0) };
> 
>         assert_eq!(result, expected_result);
> ```

I'd prefer no variable for `expected_result`.
Same below



_Suggestion:_
```Cairo
        let result = i + i;

        assert_eq!(result, CirclePoint { x: -m31(1), y: m31(0) });
```
___
*[`stwo_cairo_verifier/src/circle.cairo` line 285 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8ELUqq81uBx5xLWzO0:-O8ELUqq81uBx5xLWzO1:b88de81) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L285)):*
> ```Cairo
>             x: qm31(1, 0, 478637715, 513582971),
>             y: qm31(992285211, 649143431, 740191619, 1186584352),
>         };
> ```

This will be needed outside of tests.
Can define below `M31_CIRCLE_GEN`



_Suggestion:_
```Cairo
/// A generator for the circle group over [`SecureField`].
pub const QM31_CIRCLE_GEN: CirclePointQM31 =
    CirclePointQM31 {
        x: QM31 {
            a: CM31 { a: M31 { inner: 1 }, b: M31 { inner: 0 }, },
            b: CM31 { a: M31 { inner: 478637715 }, b: M31 { inner: 513582971 } }
        },
        y: QM31 {
            a: CM31 { a: M31 { inner: 992285211 }, b: M31 { inner: 649143431 } },
            b: CM31 { a: M31 { inner: 740191619 }, b: M31 { inner: 1186584352 } }
        },
    };
```


<!-- Sent from Reviewable.io -->

atgrosso commented:
> > Reviewed all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99)_ status: 0 of 9 files reviewed, 14 unresolved discussions (waiting on @atgrosso)
> 
> _[`stwo_cairo_verifier/src/circle.cairo` line 8 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EHdX88fWOihMDXfB3:-O8EHdX88fWOihMDXfB4:bzc5mga) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L8)):_
> 
> > ```cairo
> > pub const M31_CIRCLE_GEN: CirclePoint<M31> =
> >     CirclePoint::<M31> { x: M31 { inner: 2 }, y: M31 { inner: 1268011823 }, };
> > ```
> 
> Nit: can this be implicit
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 31 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EI3rc691JoMGgieH6:-O8EI3rd3n9PYYvNZ2Dw:brhun0a) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L31)):_
> 
> > ```cairo
> >     // Returns the neutral element of the circle.
> >     fn zero() -> CirclePoint<F> {
> >         CirclePoint::<F> { x: One::<F>::one(), y: Zero::<F>::zero() }
> > ```
> 
> Same here?
> 
> _Suggestion:_
> 
> ```cairo
> CirclePoint { x: One::one(), y: Zero::zero() }
> ```
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 50 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EIKJZ1H0k-UYe0R86:-O8EIKJ_7JgqEx-f_zRe:b-qc19fg) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L50)):_
> 
> > ```cairo
> >         };
> >         result
> >     }
> > ```
> 
> Why does this have to change so much? Can't it be the same as before just change scalar to u128?
> 
> _Code quote:_
> 
> ```cairo
>     fn mul(
>         self: @CirclePoint<F>, initial_scalar: u128
>     ) -> CirclePoint<
>         F
>     > {
>         let mut scalar = initial_scalar;
>         let mut result: CirclePoint<F> = Self::zero();
>         let mut cur: CirclePoint<F> = *self;
>         while scalar > 0 {
>             if scalar & 1 == 1 {
>                 result = result + cur;
>             }
>             cur = cur + cur;
>             scalar = scalar / 2;
>         };
>         result
>     }
> ```
> 
I removed the explicits type annotations. Regarding initial_scalar addition, i changed it to ref.

> _[`stwo_cairo_verifier/src/circle.cairo` line 54 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EIsTJ380jXrzjRHR6:-O8EIsTJ380jXrzjRHR7:b25hiq4) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L54)):_
> 
> > ```cairo
> > impl CirclePointAdd<F, +Add<F>, +Sub<F>, +Mul<F>, +Drop<F>, +Copy<F>> of Add<CirclePoint<F>> {
> >     // The operation of the circle as a group with additive notation.
> > ```
> 
> I know unrelated to your changes but mind chageing to docstring here please?
> 
> _Suggestion:_
> 
> ```cairo
>     /// Performs the operation of the circle as a group with additive notation.
> ```
> 
Changed
> _[`stwo_cairo_verifier/src/circle.cairo` line 56 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EIZVA2uexxDQRnuJc:-O8EIZVA2uexxDQRnuJd:brhun0a) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L56)):_
> 
> > ```cairo
> >     // The operation of the circle as a group with additive notation.
> >     fn add(lhs: CirclePoint<F>, rhs: CirclePoint<F>) -> CirclePoint<F> {
> >         CirclePoint::<F> { x: lhs.x * rhs.x - lhs.y * rhs.y, y: lhs.x * rhs.y + lhs.y * rhs.x }
> > ```
> 
> Same here?
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 72 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EJLXbCprTOmaef3IL:-O8EJLXbCprTOmaef3IM:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L72)):_
> 
> > ```cairo
> >         CirclePoint { x: self.x.complex_conjugate(), y: self.y.complex_conjugate() }
> >     }
> > }
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> #[generate_trait]
> pub impl ComplexConjugateImpl of ComplexConjugateTrait {
>     fn complex_conjugate(self: CirclePoint<QM31>) -> CirclePoint<QM31> {
>         CirclePoint { x: self.x.complex_conjugate(), y: self.y.complex_conjugate() }
>     }
> }
> ```
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 74 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EJiWr3isom0K8QNYo:-O8EJiWr3isom0K8QNYp:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L74)):_
> 
> > ```cairo
> > }
> > 
> > /// Represents the coset initial + <step>.
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> /// Represents the coset `initial + <step>`.
> ```
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 112 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EJunB8O86MI36Wd6Q:-O8EJunB8O86MI36Wd6R:b45h) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L112)):_
> 
> > ```cairo
> >     }
> > 
> >     fn at(self: @Coset, index: usize) -> CirclePoint::<M31> {
> > ```
> 
> _Suggestion:_
> 
> ```cairo
> CirclePoint<M31> 
> ```
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 122 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EK8E77ZTaqsFe1hxC:-O8EK8E77ZTaqsFe1hxD:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L122)):_
> 
> > ```cairo
> >     /// Creates a coset of the form G_2n + \<G_n\>.
> >     /// For example, for n=8, we get the point indices \[1,3,5,7,9,11,13,15\].
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>     /// Creates a coset of the form `G_2n + <G_n>`.
>     ///
>     /// For example, for `n=8`, we get the point indices `[1,3,5,7,9,11,13,15]`.
> ```
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 124 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EKHQsC5kuhvLI3dy1:-O8EKHQsC5kuhvLI3dy2:byp18fd) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L124)):_
> 
> > ```cairo
> >     /// For example, for n=8, we get the point indices \[1,3,5,7,9,11,13,15\].
> >     fn odds(log_size: u32) -> Coset {
> >         //CIRCLE_LOG_ORDER
> > ```
> 
> Remove
> 
Done
> _[`stwo_cairo_verifier/src/circle.cairo` line 131 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EKMkjEnCGvL1K_03V:-O8EKMkjEnCGvL1K_03W:undefined) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L131)):_
> 
> > ```cairo
> >     /// Creates a coset of the form G_4n + <G_n>.
> >     /// For example, for n=8, we get the point indices \[1,5,9,13,17,21,25,29\].
> >     /// Its conjugate will be \[3,7,11,15,19,23,27,31\].
> > ```
> 
> _Suggestion:_
> 
> ```cairo
>     /// Creates a coset of the form `G_4n + <G_n>`.
>     ///
>     /// For example, for `n=8`, we get the point indices `[1,5,9,13,17,21,25,29]`.
>     /// Its conjugate will be `[3,7,11,15,19,23,27,31]`.
> ```
> 
> _[`stwo_cairo_verifier/src/circle.cairo` line 156 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EKs_T6NmxCtSsN5zo:-O8EKs_T6NmxCtSsN5zp:b4mykql) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L156)):_
> 
> > ```cairo
> >     #[test]
> >     fn test_add_1() {
> >         let i = CirclePoint::<M31> { x: m31(0), y: m31(1) };
> > ```
> 
> Same here? and below?
> 
> _Suggestion:_
> 
> ```cairo
>         let i = CirclePoint { x: m31(0), y: m31(1) };
> ```
> 
Done! I hope i changed everyone. It seems i was a bit verbose that day 😅
> _[`stwo_cairo_verifier/src/circle.cairo` line 160 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8EL6TJ5cyGWFtC5yqk:-O8EL6TJ5cyGWFtC5yql:bdizfvm) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L160)):_
> 
> > ```cairo
> >         let expected_result = CirclePoint::<M31> { x: -m31(1), y: m31(0) };
> > 
> >         assert_eq!(result, expected_result);
> > ```
> 
> I'd prefer no variable for `expected_result`. Same below
> 
> _Suggestion:_
> 
> ```cairo
>         let result = i + i;
> 
>         assert_eq!(result, CirclePoint { x: -m31(1), y: m31(0) });
> ```
> 
Changed
> _[`stwo_cairo_verifier/src/circle.cairo` line 285 at r1](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-O8ELUqq81uBx5xLWzO0:-O8ELUqq81uBx5xLWzO1:b88de81) ([raw file](https://github.com/starkware-libs/stwo-cairo/blob/843fb386ed7997add917a5cad48528e174906975/stwo_cairo_verifier/src/circle.cairo#L285)):_
> 
> > ```cairo
> >             x: qm31(1, 0, 478637715, 513582971),
> >             y: qm31(992285211, 649143431, 740191619, 1186584352),
> >         };
> > ```
> 
> This will be needed outside of tests. Can define below `M31_CIRCLE_GEN`
> 
> _Suggestion:_
> 
> ```cairo
> /// A generator for the circle group over [`SecureField`].
> pub const QM31_CIRCLE_GEN: CirclePointQM31 =
>     CirclePointQM31 {
>         x: QM31 {
>             a: CM31 { a: M31 { inner: 1 }, b: M31 { inner: 0 }, },
>             b: CM31 { a: M31 { inner: 478637715 }, b: M31 { inner: 513582971 } }
>         },
>         y: QM31 {
>             a: CM31 { a: M31 { inner: 992285211 }, b: M31 { inner: 649143431 } },
>             b: CM31 { a: M31 { inner: 740191619 }, b: M31 { inner: 1186584352 } }
>         },
>     };
> ```
Done

andrewmilson reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/stwo-cairo/99#-:-O8wZMP66RWxKoAfYE7R:bnfp4nl)* status: 0 of 9 files reviewed, 13 unresolved discussions (waiting on @atgrosso)



<!-- Sent from Reviewable.io -->
